Clone corefx copy that matches shared framework and run tests on it.

joperezr opened the pull request:
cc: @weshaggard @ericstj

UPDATE: I have pushed more things into this PR. Basically, with the changes here we are getting the hash of the shared framework we are testing against, and cloning it into the `bin\corefxCopy` directory. After that, we are building and running the tests using the shared framework.

Note that as of now, some of the test projects are failing to build because we are now building them against the shared framework instead of netstandard.dll, but I will create separate PRs that will address them and drive our tests into a 100% success in build and pass rate.

weshaggard commented on src/SharedFrameworkValidation/RestoreSDKProject/RestoreSDKProject.csproj:
```diff
@@ -61,4 +61,14 @@
 
+  <Target Name="CopyNETCoreAppVersionsFile" DependsOnTargets="ResolveLockFileCopyLocalProjectDeps" BeforeTargets="CoreCompile">
+    <PropertyGroup>
+        <NETCoreAppVersionsFilePath Condition="$([System.String]::new('%(FileDefinitions.Identity)').Contains('Microsoft.NETCore.App.versions.txt'))">%(FileDefinitions.ResolvedPath)</NETCoreAppVersionsFilePath>
```
Won't this be potentially ambiguous if there is more then one version?

I assume you know the version and the package folder root why can't you create a direct file path?

joperezr replied:
> Won't this be potentially ambiguous if there is more then one version?

yes, it's possible that we find two, the one in the identity package and the rid specific one but either will work since corefx version is in both.

> I assume you know the version and the package folder root why can't you create a direct file path?

I don't know the package folder root, this is a SDK project that is being built using `dotnet restore` and `dotnet build` which means that the package will be restored to the `%USERPROFILE%\.nuget\packages` in the default case, but it could also be in a different place if somebody overrides the default restore path, which is why I chose to instead use the calculated one in the restored project which we know will always be correct.

weshaggard replied:
Why not just pass in the package root path to the restore commands for our testing?

joperezr replied:
AFAIK, `dotnet build` does not yet support passing in a packages directory, so basically we would fail to find the assets we restored. Would you rather me restore the package twice so that one of those times it goes in the packages folder and we can restore?

weshaggard replied:
No don't restore twice. dotnet build doesn't even support this if we use a nuget.config file to define the packages folder?

joperezr replied:
That's a good point, trying it out now.

joperezr commented:
@weshaggard your suggestion of using NuGet.config did work 👍 . I've added additional changes into this PR now that my buildtools task was merged, and updated the title of the PR with a more accurate description. PTAL: @weshaggard @ericstj

weshaggard commented on src/SharedFrameworkValidation/SharedFrameworkValidation.proj:
```diff
@@ -56,6 +56,41 @@
+    <ItemGroup>
+      <_CloneRepoLines Include="REM @echo off" />
+      <_CloneRepoLines Include="pushd $(CoreFxCopyLocation.Replace('/', '\'))" />
+      <_CloneRepoLines Include="git clone $(_RepoURL)" />
```
Wasn't the plan to just do a scoped checkout? I have a feeling if you do another clone under the bin folder which is under the outter git folder then somethings might be strange I don't know if git likes nested repo's (although I've not really tried).

joperezr replied:
I also had this question but it all seems to work just fine. I didn't do a scoped restore because @ericstj brought up a great point which is that tests might be dependent on the tools version, so the way to make sure this won't affect is by cloning again. The only possible complication this brings, is that we might hit maxPath easier, but at least in my machine this works out just fine

joperezr commented:
With the latest change, plus some changes that I need to push to fix some corefx tests, I can run this script and succesfully build all tests against the shared framework, and run all of them. I have all passing except for one test which I will investigate later today:
`System.Diagnostics.Tests.ActivityTests.IdGenerationInternalParent [FAIL]`

joperezr commented:
Merging since osx is down

karelz commented:
@joperezr this PR didn't make it in before the fork for 2.0. Is that ok? Or do we need it in rel/2.0.0 branch?

joperezr commented:
@karelz thanks for letting me know, yes I will have to merge it into rel/2.0.0. The work in this branch was about adding the infrastructure in order to test the sharedframework of the released packages before we ship. This PR only addressed the parts for the Windows validation, and I'm working for a different PR that will target non-Windows builds. Once I'm done with all of that, I'll do a cherry pick and PR against rel/2.0.0 to bring this changes in.
